xc7/utils/prjxray_assign_tile_pin_direction.py

```diff
@@ -230,45 +227,41 @@ def handle_edges_to_channels(conn, null_tile_wires, edge_assignments, channel_wi
 
 def main():
     parser = argparse.ArgumentParser()
-    parser.add_argument(
-            '--db_root', help='Project X-Ray Database', required=True)
+
     parser.add_argument(
             '--connection_database', help='Database of fabric connectivity', required=True)
     parser.add_argument(
             '--pin_assignments', help='Output JSON assigning pins to tile types and direction connections', required=True)
 
     args = parser.parse_args()
 
-    db = prjxray.db.Database(args.db_root)
-    grid = db.grid()
-
     edge_assignments = {}
-
     wires_in_tile_types = set()
 
-    for tile_type in db.get_tile_types():
-        type_obj = db.get_tile_type(tile_type)
+    with DatabaseCache(args.connection_database, read_only=True) as conn:
 
-        for wire in type_obj.get_wires():
-            wires_in_tile_types.add((tile_type, wire))
+        c = conn.cursor()
 
-        for site in type_obj.get_sites():
-            for site_pin in site.site_pins:
-                if site_pin.wire is None:
-                    continue
+        # List tile wires and site wires
+        tiles = c.execute("SELECT pkey, name FROM tile_type").fetchall()
```

Comment:
No need for the fetchall, you can just do for x in c.execute() and stream the results.  The inner for loop just needs to use a difference Cursor object.  This is recommended because it will stream the results from the database.

Reply:
Yes, but no here. The cursor "c" is used inside the loop so it will break the iteration (I did such mistakes before). I will create a second cursor eg. "c2" for iteration.